go/p2p/discovery/peerstore/peerstore.go

```diff
@@ -0,0 +1,403 @@
+package peerstore
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/common/persistent"
+	cmSync "github.com/oasisprotocol/oasis-core/go/common/sync"
+)
+
+const (
+	// PeerRegistrationTTL is a deadline for the next registration.
+	PeerRegistrationTTL = time.Hour
+
+	// storeBucketName is the name of the bucket in which backup data is stored.
+	storeBucketName = "p2p/discovery"
+
+	// storePeerstoreKey is the bucket key under which peers from the peerstore are stored.
+	storePeerstoreKey = "peerstore"
+
+	// backupStartupDelay is the startup time delay for backups.
+	backupStartupDelay = 15 * time.Minute
+
+	// backupInterval is the time interval between backups.
+	backupInterval = 15 * time.Minute
+
+	// cleanupInterval is the interval for periodic store cleanup.
+	cleanupInterval = time.Minute
+)
+
+type storeBackup struct {
+	Timestamp  time.Time         `json:"timestamp"`
+	Namespaces []namespaceBackup `json:"namespaces"`
+}
+
+type namespaceBackup struct {
+	Namespace string   `json:"namespace"`
+	Addrs     [][]byte `json:"addrs"`
+}
+
+type registration struct {
+	ns      string
+	info    peer.AddrInfo
+	expires time.Time
+
+	element *list.Element
+}
+
+// Store is an in-memory data structure for storing peers' data.
+type Store struct {
+	logger *logging.Logger
+
+	store   *persistent.CommonStore
+	storeCh chan struct{}
+
+	mu            sync.RWMutex
+	registrations map[string]map[peer.ID]*registration
+	expirations   *list.List
+
+	startOne cmSync.One
+}
+
+// NewStore creates a new peer store.
+func NewStore(cs *persistent.CommonStore) *Store {
+	return &Store{
+		logger:        logging.GetLogger("p2p/discovery/peerstore"),
+		store:         cs,
+		storeCh:       make(chan struct{}, 1),
+		registrations: make(map[string]map[peer.ID]*registration),
+		expirations:   list.New(),
+		startOne:      cmSync.NewOne(),
+	}
+}
+
+// Add inserts peer address into the store under the given namespace.
+func (s *Store) Add(ns string, info peer.AddrInfo) time.Duration {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	peers, ok := s.registrations[ns]
```

Comment:
What if a peer is submitting advertise requests and randomizing the namespace? It seems like this could cause the store to fill up unboundedly? Should there be a max cap of namespaces per peer? And a max cap of total advertisements to avoid memory exhaustion?

Also there should probably be a rate limit of advertisements per peer (probably a useful thing to have in general for our P2P RPC framework so that various methods could configure rate limits), but this is probably best done in a subsequent task.

Reply:
Rate limiting solves most of the problems, as cleanup/gc should be done periodically. Can add the cap now, and add rate limiting in a separate task.

Reply:
Yeah, for now just add the cap and we'll do rate limiting later.